packages/twenty-server/src/engine/core-modules/user-state/user-state.service.ts

```diff
@@ -0,0 +1,69 @@
+import { Injectable } from '@nestjs/common';
+
+import { SkipSyncEmail } from 'src/engine/core-modules/user-state/dtos/skip-sync-email.entity';
+import { InjectObjectMetadataRepository } from 'src/engine/object-metadata-repository/object-metadata-repository.decorator';
+import { ConnectedAccountWorkspaceEntity } from 'src/modules/connected-account/standard-objects/connected-account.workspace-entity';
+import { ConnectedAccountRepository } from 'src/modules/connected-account/repositories/connected-account.repository';
+import { User } from 'src/engine/core-modules/user/user.entity';
+import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
+import { UserState } from 'src/engine/core-modules/user-state/dtos/user-state.dto';
+import {
+  KeyValuePairService,
+  KeyValueTypes,
+} from 'src/engine/core-modules/key-value-pair/key-value-pair.service';
+import { UserStates } from 'src/engine/core-modules/user-state/enums/user-states.enum';
+import { UserStateEmailSyncValues } from 'src/engine/core-modules/user-state/enums/user-state-email-sync-values.enum';
+
+@Injectable()
+export class UserStateService {
+  constructor(
+    @InjectObjectMetadataRepository(ConnectedAccountWorkspaceEntity)
+    private readonly connectedAccountRepository: ConnectedAccountRepository,
+    private readonly keyValuePairService: KeyValuePairService<KeyValueTypes.USER_STATE>,
+  ) {}
+
+  async getUserState(user: User, workspace: Workspace): Promise<UserState> {
+    if (!user || !workspace) {
+      return {
+        skipSyncEmail: true,
+      };
+    }
+    const connectedAccounts =
+      await this.connectedAccountRepository.getAllByUserId(
+        user.id,
+        workspace.id,
+      );
+
+    if (connectedAccounts?.length) {
+      return {
+        skipSyncEmail: true,
```

Comment:
This could be misleading too. Having an account does not mean we skipped the sync (it probably means the opposite actually). I know why you are doing this here but I'm a bit confused with this UserState overall. Maybe it will get clearer with more info in that object but in this specific feature I would have imagined an onboarding step field like a state machine where you store your current step from step A to step B.
cc @charlesBochet probably approved this design and has more context than me but still 🙂

Reply:
yeah, i understand why you are confused, this `skipSyncEmail` state in indead `skipSyncEmailOnboardingStep`, which should be true if we already have a connectedAccount OR if we already skipped this step
I will rename that field `skipSyncEmailOnboardingStep` would that be clearer for you?

Reply:
Since this user state can grow in size with time considering the data structure you chose, I feel like adding more context in the name would help yes! @martmull 🙂